pkg/k8s/rebalance.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	log "github.com/Sirupsen/logrus"
+
+	"context"
+	"time"
+
+	"github.com/projectcalico/typha/pkg/config"
+)
+
+type MaxConnsAPI interface {
+	SetMaxConns(numConns int)
+}
+
+type K8sAPI interface {
+	GetNumTyphas(namespace, serviceName, portName string) (int, error)
+	GetNumNodes() (int, error)
+}
+
+func PollK8sForConnectionLimit(
+	cxt context.Context,
+	configParams *config.Config,
+	tickerC <-chan time.Time,
+	k8sAPI K8sAPI,
+	server MaxConnsAPI,
+) {
+	logCxt := log.WithField("thread", "k8s-poll")
+	logCxt.Info("Kubernetes poll goroutine started.")
+	activeTarget := configParams.MaxConnectionsUpperLimit
+	for {
+		select {
+		case <-tickerC:
+			// Get the number of Typhas in the service.
+			numTyphas, tErr := k8sAPI.GetNumTyphas(
+				configParams.K8sNamespace,
+				configParams.K8sServiceName,
+				configParams.K8sPortName,
+			)
+			if tErr != nil || numTyphas <= 0 {
+				logCxt.WithError(tErr).Warn("Failed to get number of Typhas")
+			}
+			// Get the number of nodes as an estimate for the number of Felix connections we should expect.
+			numNodes, nErr := k8sAPI.GetNumNodes()
+			if nErr != nil || numNodes <= 0 {
+				logCxt.WithError(nErr).Warn("Failed to get number of nodes")
```

Comment:
I'm being very defensive in this code and trying to sanity check everything we're getting.  Feels worth it to avoid shooting ourselves in the foot!